src/mame/misc/anes.cpp

```diff
@@ -38,138 +39,447 @@ namespace {
+	else        { y0 = 0;       y1 = sh;    dy = +1; }
+
+	int addr = src;
+	for (int y = y0; y != y1; y += dy)
+	{
+		for (int x = x0; x != x1; x += dx)
+		{
+			int pen = (which & 0x10) ? 0 : m_blitrom[(addr++) % m_blitrom.bytes()];
+
+//          if (pen != 0xff)
+				bitmap.pix((sy + y) & 0x1ff, (sx + x) & 0x1ff) = pen;
+		}
+	}
+}
+
+void anes_state::rombank_w(uint8_t data)
+{
+	if (data & (~0x07))
+		logerror("%s: Unknown ROM bank bit written %02X\n", machine().describe_context(), data);
+
+	m_bank = data & 0x07;
+
+	// the bank change must happen AFTER the next instruction, be it a jmp or a ret
+	m_bank_delay = 2;
+}
+
+void anes_state::blit_w(offs_t offset, uint8_t data)
+{
+	m_blit[offset] = data;
+	if (offset == 0x0a)
+		do_blit();
+}
+
+void anes_state::blit_rom_w(offs_t offset, uint8_t data)
+{
+	uint8_t which = m_blit[0x0b];
+//  logerror("%s: blit ROM written %02X (%04X <- %02X)\n", machine().describe_context(), which, offset, data);
+	if (which == 1)
+	{
+		m_blit_addr[0] = offset;
+		m_blit_val[0] = data;
+	}
+	else if (which == 2)
+	{
+		m_blit_addr[1] = offset;
+		m_blit_val[1] = data;
+	}
+	else
+		logerror("%s: Unknown blit ROM which %02X\n", machine().describe_context(), which);
+}
+
+uint8_t anes_state::blit_status_r()
+{
+	// bits 0-3 = ?
+	// bit    4 = busy
+	return 0x00;
+}
+
+void anes_state::mux_w(uint8_t data)
 {
-	m_vram_offset[offset] = data;
+	m_mux = data;
+	if (m_mux & (~0x1f))
+		logerror("%s: Unknown mux bit written %02X\n", machine().describe_context(), data);
 }
 
-void anes_state::blit_trigger_w(uint8_t data)
+uint8_t anes_state::key_r(offs_t offset)
 {
-	/*
-	 operation is:
-	 checks for bit 4 to be high in port $16
-	 writes either 0 or 3 to port $04
-	 writes a 0 to port $00
-	 writes an offset to ports $0c / $0d / $0e
-	 writes 1 to port $0b, writes to program space, writes 2 to port $0b, writes to program space
-	 writes a mode to port $0b, writes to trigger port $0a
-	 */
+	for (int i = 0; i < 5; i++)
+		if (!BIT(m_mux, i))
+			return m_key[offset][i]->read();
 
-	//printf("%02x%02x%02x\n",m_vram_offset[0],m_vram_offset[1],m_vram_offset[2]);
+	//  logerror("%s: Unknown key read with mux = %02X\n", machine().describe_context(), m_mux);
+	//  return 0xff;
+	return offset ? m_coin->read() : 0xff;
+}
+
+
+uint8_t anes_state::fixed_rom_r(offs_t offset)
+{
+	uint8_t ret = m_maincpu->space(AS_PROGRAM).read_byte(offset);
+	if (!machine().side_effects_disabled())
+	{
+		if (m_bank_delay)
+		{
+			logerror("reading non-banked rom area with bank delay %04x: %02x\n", offset, ret);
+			m_banktimer->adjust(attotime::from_usec(1));
+		}
+	}
+	return ret;
+}
+
+uint8_t anes_state::banked_rom_r(offs_t offset)
+{
+	uint8_t ret = m_maincpu->space(AS_PROGRAM).read_byte(offset + 0x2000);
+	if (!machine().side_effects_disabled())
+	{
+		if (m_bank_delay)
+		{
+			logerror("reading banked rom area with bank delay %04x: %02x\n", offset + 0x2000, ret);
+			m_banktimer->adjust(attotime::from_usec(1));
```

Comment:
Note: smc777.cpp uses Z80 R register because it's more reliable than using a timer. 
Even better, it's probably possible to use Z80 `refresh_cb` for both nowadays, likely this uses M1 cycles too.

Reply:
I think poking directly into CPU registers is a pretty gross hack compared to a timer...

Reply:
well M1 stuff is what MAME does through AS_OPCODES, which is where the code to start the timer is hooked up.

I don't know what refresh_cb is in this context, does it happen after an opcode?

Reply:
https://github.com/mamedev/mame/blob/8f3716ad43b45fbef9cb35e20bc84c078a727c60/src/devices/cpu/z80/z80.cpp#L522

z80 core reads the opcode, sends the callback to client finally updates the internal state.

Reply:
refresh_cb is preliminary emulation of RFSH pin.
AS_OPCODES is (a weird way instead of a cb) for M1. Though I'd use this one since it's more likely how it works on hardware. And instead of a timer, add a counter? This is also more likely from hw point of view. A 1usec timer will have problems when/if the z80 core emulation gets more accurate.

Reply:
I don't really see how a counter can work when it must execute after a single byte 'ret' or a 3 byte 'jmp'

Reply:
because M1 is triggered once per opcode, unless you use prefixed opcodes.

Reply:
but doesn't it happen when the first byte is fetched? meaning it'll fetch the wrong address on a jmp if we switch straight away, or fetch 2 wrong opcodes after a ret if we force a counter based delay?

Reply:
Ah, AS_OPCODES is for opcode arg too? Then yes I'm wrong and it means our Z80 core doesn't emulate the M1 pin.

Reply:
Delayed bankswitch can be done with m1 counting:
https://github.com/mamedev/mame/blob/8f3716ad43b45fbef9cb35e20bc84c078a727c60/src/mame/sinclair/pentevo.cpp#L671

Reply:
so can somebody just give me a (clean) solution here that will work for THIS case.

1. bankswitch value is written to port 0xfe
2. next opcode must be fetched and execute in full, including the oprand if present ('ret' or 'jmp' or 'call')
3. bank must change (so code after ret / jmp / call has been executed is in the new bank)

I volunteered to look at this because I wanted to concentrate on the video emulation, not this, this is putting me off.

Reply:
operands don't go through AS_OPCODES.

Reply:
I removed my local anes.cpp so I can't check, but yeah I can see why the debugger will go wrong.

now:
```
debugger_instruction_hook(PCD);
uint8_t opcode = rop();
````

should be:
```
assert M1
callback to anes bankswitch here
debugger_instruction_hook(PCD);
uint8_t opcode = rop();
clear M1
````

Reply:
yes, oprands don't go through AS_OPCODES
but the opcode comes first, I don't see how the logic you've presented would know to not switch the bank before the oprands have been fetched, or in the case of a ret, know that none are needed?